### ISSUE TITLE
Does opam support user patches / overriding package files?

### ISSUE THREAD

**Issue by hyphenrf** [ocaml/opam#5020]
Hello, after going through the manual and a couple `--help` menus, I'm not sure what mechanism opam exposes for overriding an opam file in a remote package.
Since opam is a source-based package manager, I imagine there exists some sort of mechanism for storing and invoking local patches a la portage, or for overriding a remote package with a local version. Help?

**Comment by kit-ty-kate** [comment]
What do you want to do exactly? Would `opam pin` be enough?

**Comment by MSoegtropIMC** [comment]
For overriding remote packages with local packages, you can create a local opam repo (essentially a folder structure) and register it with a higher priority than the remote repo. You can find an elaborated example here:

https://github.com/coq/platform/blob/7cf8bc1bc063d60d3be422489931295c523e8b1a/shell_scripts/install_opam.sh#L184

https://github.com/coq/platform/tree/main/opam

For overriding sources for a specific package, you can add patch files, as e.g. in:

https://github.com/coq/platform/tree/main/opam/opam-repository/packages/lablgtk3/lablgtk3.3.1.1

https://github.com/coq/platform/blob/7cf8bc1bc063d60d3be422489931295c523e8b1a/opam/opam-repository/packages/lablgtk3/lablgtk3.3.1.1/opam#L26

**Comment by hyphenrf** [comment]
> What do you want to do exactly?

Two separate things came up that made me interested in such mechanism:
- I have a library installed from the default opam repo, and I made some patches to it that I want to apply automatically whenever opam updates it.
- Few ocaml releases back I was working on multicore build process and I was thinking "there must be a faster feedback loop than initializing my own repo that pulls from my github remote". I was looking for a way to simply override an opam file of an existing package in a remote repo and use that local override whenever I'm querying/installing/etc.. the package.

**Comment by hyphenrf** [comment]
@MSoegtropIMC thanks for the reference files, checking them out rn

**Comment by kit-ty-kate** [comment]
> * I have a library installed from the default opam repo, and I made some patches to it that I want to apply automatically whenever opam updates it.

There isn’t any feature in opam for that that I know of. You can have a local pin to the package with your patch or create a custom repository but in both cases the patches will not automagically get moved to the next version if updated sadly.
I’ll add the feature request label to this issue then.

> * Few ocaml releases back I was working on multicore build process and I was thinking "there must be a faster feedback loop than initializing my own repo that pulls from my github remote". I was looking for a way to simply override an opam file of an existing package in a remote repo and use that local override whenever I'm querying/installing/etc.. the package.

`opam pin edit <the package you want to modify>` should be what you are looking for, for that, if i understand correctly.

**Comment by hyphenrf** [comment]
yeah a local repo will do fine I suppose. It seems the mechanism for automatic patching exists, just that the way to do it at user level seems a little involved as it requires more-or-less full maintenance of the local opam files (and subsequently keeping in-sync with remote updates).

for completeness, the idea comparable to this feature-wish is in portage: https://wiki.gentoo.org/wiki//etc/portage/patches

**Comment by MSoegtropIMC** [comment]
Indeed for each new remote version you have to manually create a local opam package with the patch. It is not that hard, though, to automate it. You do an opam update, run a script which checks each package you have locally for updates and if there is one patch the update (diff the previous opam package and apply the patch to the new one) and run opam update again. A sort of meta patch.

Another method might be to do a local opam package with a high version which points to e.g. a master branch and update the package at appropriate times.
